core/src/layers/prometheus.rs

```diff
@@ -93,32 +89,117 @@ use crate::*;
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use log::debug;
+    /// # use opendal::layers::PrometheusLayer;
+    /// # use opendal::services;
+    /// # use opendal::Operator;
+    /// # use opendal::Result;
+    /// #
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<()> {
+    ///     // Pick a builder and configure it.
+    ///     let builder = services::Memory::default();
+    ///     let registry = prometheus::default_registry();
+    ///
+    ///     let op = Operator::new(builder)
+    ///         .expect("must init")
+    ///         .layer(PrometheusLayer::new(registry))
+    ///         .finish();
+    ///     debug!("operator: {op:?}");
+    ///
+    ///     Ok(())
+    /// # }
+    /// ```
+    pub fn new(registry: &Registry) -> Self {
+        PrometheusLayerBuilder::default().register(registry)
+    }
+
+    /// Create a [`PrometheusLayerBuilder`] to modify the default metric configuration.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use log::debug;
+    /// # use opendal::layers::PrometheusLayer;
+    /// # use opendal::services;
+    /// # use opendal::Operator;
+    /// # use opendal::Result;
+    /// #
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<()> {
+    ///     // Pick a builder and configure it.
+    ///     let builder = services::Memory::default();
+    ///     let registry = prometheus::default_registry();
+    ///
+    ///     let duration_seconds_buckets = prometheus::exponential_buckets(0.01, 2.0, 16).unwrap();
+    ///     let bytes_buckets = prometheus::exponential_buckets(1.0, 2.0, 16).unwrap();
+    ///     let op = Operator::new(builder)
+    ///         .expect("must init")
+    ///         .layer(
+    ///             PrometheusLayer::builder()
+    ///                 .operation_duration_seconds_buckets(duration_seconds_buckets)
+    ///                 .operation_bytes_buckets(bytes_buckets)
+    ///                 .enable_path_label(1)
+    ///                 .register(registry)
+    ///         )
+    ///         .finish();
+    ///     debug!("operator: {op:?}");
+    ///
+    ///     Ok(())
+    /// # }
+    /// ```
+    pub fn builder() -> PrometheusLayerBuilder {
+        PrometheusLayerBuilder::default()
+    }
+}
+
+impl<A: Access> Layer<A> for PrometheusLayer {
+    type LayeredAccess = observe::MetricsAccessor<A, PrometheusInterceptor>;
+
+    fn layer(&self, inner: A) -> Self::LayeredAccess {
+        observe::MetricsLayer::new(self.interceptor.clone()).layer(inner)
+    }
+}
+
+/// [`PrometheusLayerBuilder`] is a config builder to build a [`PrometheusLayer`].
+pub struct PrometheusLayerBuilder {
     operation_duration_seconds_buckets: Vec<f64>,
     operation_bytes_buckets: Vec<f64>,
     path_label_level: usize,
 }
 
-impl PrometheusLayer {
-    /// Create a [`PrometheusLayer`] while registering itself to this registry.
-    pub fn new(registry: Registry) -> Self {
+impl Default for PrometheusLayerBuilder {
```

Comment:
How about removing `Default` from `PrometheusLayerBuilder`? We can provide a private `new` function for `PrometheusLayerBuilder` to prevent users from creating a builder outside the layer.

Reply:
We have not exported the `PrometheusLayerBuilder` type, so I think implementing `Default` might be more consistent with the current semantics (default config of metrics)?

Reply:
`PrometheusLayerBuilder` should be exposed; otherwise, users can't access its API documentation.